backend/filen/filen.go

```diff
@@ -0,0 +1,1084 @@
+}
+
+// Directory is Filen's directory type
+type Directory struct {
+	fs        *Fs
+	path      string
+	directory types.DirectoryInterface
+}
+
+// Fs returns read only access to the Fs that this object is part of
+func (dir *Directory) Fs() fs.Info {
+	return dir.fs
+}
+
+// String returns a description of the Object
+func (dir *Directory) String() string {
+	if dir == nil {
+		return "<nil>"
+	}
+	return dir.Remote()
+}
+
+// Remote returns the remote path
+func (dir *Directory) Remote() string {
+	return dir.fs.Enc.ToStandardPath(dir.path)
+}
+
+// ModTime returns the modification date of the file
+// It should return a best guess if one isn't available
+func (dir *Directory) ModTime(ctx context.Context) time.Time {
+	directory, ok := dir.directory.(*types.Directory)
+	if !ok {
+		return time.Time{} // todo add account creation time?
+	}
+
+	if directory.Created.IsZero() {
+		obj, err := dir.fs.filen.FindDirectory(ctx, dir.fs.resolvePath(dir.path))
+		newDir, ok := obj.(*types.Directory)
+		if err != nil || !ok {
+			return time.Now()
+		}
+		directory = newDir
+		dir.directory = newDir
+	}
+	return directory.Created
+}
+
+// Size returns the size of the file
+//
+// filen doesn't have an efficient way to find the size of a directory
+func (dir *Directory) Size() int64 {
+	return -1
+}
+
+// Items returns the count of items in this directory or this
+// directory and subdirectories if known, -1 for unknown
+func (dir *Directory) Items() int64 {
+	return -1
+}
+
+// ID returns the internal ID of this directory if known, or
+// "" otherwise
+func (dir *Directory) ID() string {
+	return dir.directory.GetUUID()
+}
+
+// File is Filen's normal file
+type File struct {
+	fs   *Fs
+	path string
+	file *types.File
+}
+
+// Fs returns read only access to the Fs that this object is part of
+func (file *File) Fs() fs.Info {
+	return file.fs
+}
+
+// String returns a description of the Object
+func (file *File) String() string {
+	if file == nil {
+		return "<nil>"
+	}
+	return file.Remote()
+}
+
+// Remote returns the remote path
+func (file *File) Remote() string {
+	return file.fs.Enc.ToStandardPath(file.path)
+}
+
+// ModTime returns the modification date of the file
+// It should return a best guess if one isn't available
+func (file *File) ModTime(ctx context.Context) time.Time {
+	if file.file.LastModified.IsZero() {
+		newFile, err := file.fs.filen.FindFile(ctx, file.fs.resolvePath(file.path))
+		if err == nil && newFile != nil {
+			file.file = newFile
+		}
+	}
+	return file.file.LastModified
+}
+
+// Size returns the size of the file
+func (file *File) Size() int64 {
+	return int64(file.file.Size)
+}
+
+// Hash returns the selected checksum of the file
+// If no checksum is available it returns ""
+func (file *File) Hash(ctx context.Context, ty hash.Type) (string, error) {
+	if ty != hash.SHA512 {
+		return "", hash.ErrUnsupported
+	}
+	if file.file.Hash == "" {
+		foundFile, err := file.fs.filen.FindFile(ctx, file.fs.resolvePath(file.path))
```

Comment:
Will reading the Hash take another transaction normally? Or does the Hash arrive in the directory listings?

If it always takes another transaction then you should set the `SlowHash` feature flag.

Reply:
Hash arrives in directory listings, it's part of the encrypted metadata so anytime we get file info we also get the hash